### PR TITLE
feat: handle parentheses with unary ops

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

The one last piece

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced a new `matches` function for advanced pattern matching in SQL queries.

- **Bug Fixes**
  - Enhanced parsing logic to handle unary operators (`Must`, `Negative`, `Optional`) more effectively in pattern matching.

- **Tests**
  - Added new SQL test cases to demonstrate and verify the usage of the `matches` function.

### CODE REVIEW

**Comment by waynexia** [comment]
@coderabbitai pause

File: src/common/function/src/scalars/matches.rs [hunk 1]
```diff
@@ -1138,11 +1264,27 @@ mod test {
...
-            //     "(+fox +jumps) over",
-            //     vec![true, true, true, true, true, true, true],
-            // ),
+            // weird parentheses cases
+            (
+                "(+fox +jumps) over",
+                vec![true, true, true, true, true, true, true],
+            ),
+            (
+                "+(fox jumps) AND over",
```
Comment (zhongzc) [inline src/common/function/src/scalars/matches.rs:1273]: Test name should be `evaluate_matches`

File: tests/cases/standalone/common/select/matches.sql [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+insert into fox values
+    (1, 'The quick brown fox jumps over the lazy dog'),
+    (2, 'The             fox jumps over the lazy dog'),
+    (3, 'The quick brown     jumps over the lazy dog'),
+    (4, 'The quick brown fox       over the lazy dog'),
+    (5, 'The quick brown fox jumps      the lazy dog'),
+    (6, 'The quick brown fox jumps over          dog'),
+    (7, 'The quick brown fox jumps over the      dog');
+
+select fox from fox where matches(fox, '"fox jumps"') order by ts;
```
Comment (zhongzc) [inline tests/cases/standalone/common/select/matches.sql:15]: Can you add cases shown in `evaluate_matches`?

Comment (killme2008) [reply tests/cases/standalone/common/select/matches.sql:15]: +1

File: src/common/function/src/scalars/matches.rs [hunk 3]
```diff
@@ -1138,11 +1264,27 @@ mod test {
...
+                vec![true, true, true, true, false, true, true],
```
Comment (killme2008) [inline src/common/function/src/scalars/matches.rs:1274]: Need to add some bad cases such as missing left paren or right paren like that.

**Review by zhongzc — APPROVED** [review]

File: src/common/function/src/scalars/matches.rs [hunk 4]
```diff
@@ -980,18 +1136,21 @@ mod test {
...
-                                    op: UnaryOp::Optional,
-                                    pattern: "b".to_string(),
-                                },
-                                PatternAst::Literal {
-                                    op: UnaryOp::Optional,
-                                    pattern: "c".to_string(),
-                                },
-                            ],
+                        PatternAst::Group {
+                            op: UnaryOp::Optional,
```
Comment (evenyag) [inline src/common/function/src/scalars/matches.rs:1140]: Can we add some tests for groups with `Must` or `Negative` op? Now we only have tests for groups with `Optional` op.

**Review by evenyag — APPROVED** [review]
Others LGTM

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/common/function/src/scalars/matches.rs (hunk 4 above)
Comment (waynexia) [reply src/common/function/src/scalars/matches.rs:1140]: `evaluate_matches` has several cases like `+(fox jumps) AND over` or `over -(fox jumps)`
